### PR TITLE
WIP Improvement: Introduce IRenderService with IBitmapRegistry Instance

### PR DESCRIPTION
**Pull request by inforithmics**
Things done:
1. Introduce a IRenderService Interface
2. Introduce a IBitmapRegistry Instance per RenderService so that registered images during Rendering are disposed along with the IRenderService.
3. Made BitmapRegistry Thread Safe.
4. Map BitmapRegistry Hierarchical so that Symbols that are registered in BitmapRegistry.Instance are accessible for the Child BitmapRegistry. The BitmapIds are globally unique.
5. Implements this https://github.com/Mapsui/Mapsui/issues/1998

Possible Improvements the IRenderService could be enhanced to Create HttpClients for Fetching Data, I didn't implement here anything like that because this could cause conflicts with a possible rewrite of a central data fetcher.

### CODE REVIEW

**Review by pauldendulk — COMMENTED** [review]
I really like the idea of a RenderService to hold all the state needed for rendering. I also like the BitmapRegistry as service a lot. I currently have two objections:
- The object Bitmap is something we need to get rid of but this PR also adds it to the SymbolStyle.
- The BitmapRegistry is now still a half singleton. It should be an instance only. To make this possible however we need to access this instance in all places were we register a bitmap. Often that is on the users side of things, far away from the renderer. This would become complicated. The solution would be to implement this first: https://github.com/Mapsui/Mapsui/issues/1998

**Comment by pauldendulk** [comment]
> Introduce a IBitmapRegistry Instance per RenderService so that registered images during Rendering are disposed along with the IRenderService.

Could you explain a bit more about the responsibilities? Who should dispose what? Should it be possible to share a cache between renderers?

**Comment by inforithmics** [comment]
Good idea with the IconPath. This would solve rendering sprites without always loading the bitmap.

Disposing responsability. Currently the Cache and Render Service are a members of the Control and when the control owns the cache and services are disposed with the Control. But I have to check if it is really the case.

File: Mapsui/Styles/SymbolStyle.cs [hunk 1]
```diff
@@ -43,6 +45,23 @@ public int BitmapId
         }
     }
 
+    /// <summary>
+    /// Bitmap of the image to display during rendering a BitmapId is assigned.
+    /// </summary>
+    public object? Bitmap
```
[the remark this answers is not in the document]
Comment (inforithmics) [reply Mapsui/Styles/SymbolStyle.cs:51]: I'll use the IconPath here so this avoids disposable objects in the style

File: Mapsui/Styles/BitmapRegistry.cs [hunk 2]
```diff
@@ -1,18 +1,27 @@
...
     private static BitmapRegistry? _instance;
-    private readonly Dictionary<int, object> _register = [];
-    private readonly Dictionary<string, int> _lookup = [];
+    private readonly ConcurrentDictionary<int, object> _register = [];
+    private readonly ConcurrentDictionary<string, int> _lookup = [];
+    private readonly IBitmapRegistry? _parent;
     private BitmapRegistry() { }
-    private int _counter = 1;
+
+    public BitmapRegistry(IBitmapRegistry parent)
```
[the remark this answers is not in the document]
Comment (inforithmics) [reply Mapsui/Styles/BitmapRegistry.cs:19]: I think with the IconPath there isn't any need for a static instance of BitmapRegistry.

File: Mapsui/Rendering/IRenderService.cs [hunk 3]
```diff
@@ -0,0 +1,8 @@
+﻿using Mapsui.Styles;
+
+namespace Mapsui.Rendering;
+
+public interface IRenderService : IRenderCache, IBitmapRegistry
```
[the remark this answers is not in the document]
Comment (inforithmics) [reply Mapsui/Rendering/IRenderService.cs:5]: Good idea, I'll takle this when Iconpath is implemented

[the remark this answers is not in the document]
Comment (inforithmics) [reply Mapsui/Rendering/IRenderService.cs]: Is now done

File: Mapsui.Rendering.Skia/SkiaStyles/SymbolStyleRenderer.cs [hunk 4]
```diff
@@ -124,9 +124,13 @@ internal static async void LoadBitmapId(SymbolStyle symbolStyle, IBitmapRegistry
         try
         {
             if (symbolStyle.Bitmap != null)
+            {
```
Comment (pauldendulk) [inline Mapsui.Rendering.Skia/SkiaStyles/SymbolStyleRenderer.cs:127]: I prefer to remove unnecessary braces in the case of single line calls. There is no general consensus on this but I think there is a general trend toward more compact code and I like that. One thing I am not certain about is if-then constructs on one line. I do that a lot too but perhaps this should be on two lines. Perhaps we can add that to the .editorconfig.

File: Mapsui/Styles/SymbolStyle.cs [hunk 5]
```diff
@@ -55,7 +55,7 @@ public int BitmapId
             if (value != null)
             {
                 SymbolType = SymbolType.Image;
-                _bitmap = null;
+                _bitmapPath = null;
```
Comment (pauldendulk) [inline Mapsui/Styles/SymbolStyle.cs:58]: oh, here you use a _bitmapPath, that is good. Perhaps I was looking at old commits before.

File: Mapsui/Styles/BitmapRegistry.cs [hunk 6]
```diff
@@ -25,7 +25,7 @@ public BitmapRegistry(IBitmapRegistry parent)
         _parent = parent;
     }
 
-    private static int _counter = 0;
+    private int _counter;
```
Comment (pauldendulk) [inline Mapsui/Styles/BitmapRegistry.cs:30]: 👍

**Review by pauldendulk — COMMENTED** [review]
I added some more comments. I see you are incorporating my remarks, great, yet there is a problem with how this MR grows and how every change, although mostly an improvement, is a reason for more comments. The main problem here is that the basis of the current bitmap registry needs improvement. So the problem is not in this MR but in the current state of the code. Fixes to the basis should be done in separate merge requests and those can be merged later into this MR. Also I must say that as a reviewer I lose track of exactly which problem we are solving in such a growing PR. Not sure how to deal with this. I need to play around with the code myself. Hopefully I will have some time for this today.

**Comment by pauldendulk** [comment]
> I added some more comments. I see you are incorporating my remarks, great, yet there is a problem with how this MR grows and how every change, although mostly an improvement, is a reason for more comments. The main problem here is that the basis of the current bitmap registry needs improvement. So the problem is not in this MR but in the current state of the code. Fixes to the basis should be done in separate merge requests and those can be merged later into this MR. Also I must say that as a reviewer I lose track of exactly which problem we are solving in such a growing PR. Not sure how to deal with this. I need to play around with the code myself. Hopefully I will have some time for this today.

@inforithmics I did a small MR adding the IRenderService, but there is a lot more involved in this PR. Too much for me to properly review. This is primarily a problem because of the code it builds on. So I would like to ask if you could add the url based bitmap caching code in a separate PR, so I could look at that first.

**Comment by inforithmics** [comment]
> > I added some more comments. I see you are incorporating my remarks, great, yet there is a problem with how this MR grows and how every change, although mostly an improvement, is a reason for more comments. The main problem here is that the basis of the current bitmap registry needs improvement. So the problem is not in this MR but in the current state of the code. Fixes to the basis should be done in separate merge requests and those can be merged later into this MR. Also I must say that as a reviewer I lose track of exactly which problem we are solving in such a growing PR. Not sure how to deal with this. I need to play around with the code myself. Hopefully I will have some time for this today.
> 
> @inforithmics I did a small MR adding the IRenderService, but there is a lot more involved in this PR. Too much for me to properly review. This is primarily a problem because of the code it builds on. So I would like to ask if you could add the url based bitmap caching code in a separate PR, so I could look at that first.

Certainly I will do a separate pull request for that.

**Comment by inforithmics** [comment]
Closing this now, will do smaller PRs instead

**Pull request closed without merging**